src/bitcoin-cli.cpp

```diff
@@ -569,6 +569,56 @@ class NetinfoRequestHandler : public BaseRequestHandler
         "> bitcoin-cli -netinfo help\n"};
 };
 
+/** Process addressinfo requests */
+class AddressinfoRequestHandler : public BaseRequestHandler
+{
+private:
+    static constexpr uint8_t m_networks_size{5};
+    const std::array<std::string, m_networks_size> m_networks{{"ipv4", "ipv6", "torv2", "torv3", "i2p"}};
+    std::array<uint64_t, m_networks_size> m_counts{{}}; // Address counts by network
+    uint64_t m_total{0}; // Total address count
+    int8_t NetworkStringToId(const std::string& str) const
+    {
+        for (uint8_t i = 0; i < m_networks_size; ++i) {
+            if (str == m_networks.at(i)) return i;
+        }
+        return UNKNOWN_NETWORK;
+    }
+
+public:
+    UniValue PrepareRequest(const std::string& method, const std::vector<std::string>& args) override
+    {
+        if (!args.empty()) {
+            throw std::runtime_error("-addressinfo takes no arguments");
+        }
+        UniValue params{RPCConvertValues("getnodeaddresses", std::vector<std::string>{{"0"}})};
+        return JSONRPCRequestObj("getnodeaddresses", params, 1);
+    }
+
+    UniValue ProcessReply(const UniValue& reply) override
+    {
+        // Count the number of peers we know, by network.
+        for (const UniValue& node : reply["result"].getValues()) {
+            std::string network_name{node["network"].get_str()};
+            if (network_name == "onion") {
+                network_name = node["address"].get_str().size() > 22 ? "torv3" : "torv2";
+            }
+            const int8_t network_id{NetworkStringToId(network_name)};
+            if (network_id == UNKNOWN_NETWORK) continue;
+            ++m_counts.at(network_id);
+        }
+        // Prepare result to return to user.
+        UniValue result{UniValue::VOBJ}, addresses{UniValue::VOBJ};
+        for (uint8_t i = 0; i < m_networks_size; ++i) {
+            addresses.pushKV(m_networks.at(i), m_counts.at(i));
+            m_total += m_counts.at(i);
+        }
+        addresses.pushKV("total", m_total);
+        result.pushKV("addresses known", addresses);
```

Comment:
57eb71ce0f27c0353ea0742a696c6d04ce048e15

IMO avoid spaces in keys.

Reply:
Good point, will snakecase it if this remains in JSON format. I wonder if it should drop the JSON format (like `-netinfo`) for the client side.

Reply:
done